src/index.d.ts

```diff
@@ -0,0 +1,158 @@
+
+  interface FormMeta {
+    error: string | any
+    isSubmitting: boolean
+    isDirty: boolean
+    isSubmitted: boolean
+    submissionAttempts: number
+    isValid: boolean
+    fieldsAreValidating: boolean
+    fieldsAreValid: boolean
+    canSubmit: boolean
+    [k: string]: any
+  }
+
+  type SetFormMeta = Partial<FormMeta> | ((previousMeta: FormMeta) => FormMeta)
+
+  interface FormInstance<T> {
+    Form: ComponentType<Omit<HTMLProps<HTMLFormElement>, 'onSubmit'>>
+    values: T
+    meta: FormMeta
+    formContext: FormInstance<T>
+    reset: () => void
+    setMeta: (value: SetFormMeta) => void
+    handleSubmit: FormEventHandler<HTMLFormElement>
+    debounce: Debounce
+    setValues: Dispatch<SetStateAction<T>>
+    runValidation: () => void
+    getFieldValue: <V = any>(fieldPath: keyof T | string) => V
+    getFieldMeta: (
+      fieldPath: keyof T | string
+    ) => { error: string | null; [k: string]: any }
+    setFieldValue: <V = any>(
+      fieldPath: keyof T | string,
+      updater: SetValue<V>,
+      options?: { isTouched: boolean }
+    ) => void
+    setFieldMeta: (fieldPath: keyof T | string, value: SetFieldMeta) => void
+    pushFieldValue: <V = any>(fieldPath: keyof T | string, value: V) => void
+    insertFieldValue: <V = any>(
+      fieldPath: keyof T | string,
+      insertIndex: number,
+      value: V
+    ) => void
+    removeFieldValue: (
+      fieldPath: keyof T | string,
+      removalIndex: number
+    ) => void
+    swapFieldValues: (
+      fieldPath: keyof T | string,
+      firstIndex: number,
+      secondIndex: number
+    ) => void
+  }
+
+  interface FormOptions<T> {
+    defaultValues: T
+    onSubmit: (values: T, instance: FormInstance<T>) => OptionalPromise<void>
+    validate: (
+      values: Partial<T>,
+      instance: FormInstance<T>
+    ) => OptionalPromise<ValidatorReturn>
+    validatePristine?: boolean
+    debugForm: boolean
+  }
+
+  export function useForm<T extends {} = {}>(
+    options?: Partial<FormOptions<T>>
+  ): FormInstance<T>
+
+  export function useFormContext<T extends {} = {}>(): FormInstance<T>
+
+  interface FieldMeta {
+    error: string | false
+    isTouched: boolean
+    [k: string]: any
+  }
+
+  type SetValue<S> = S | ((prevState: S) => S)
+  type SetFieldMeta =
+    | Partial<FieldMeta>
+    | ((previousMeta: FieldMeta) => FieldMeta)
+
+  interface FieldInstance<T = any, F = any> {
+    form: FormInstance<F>
+    fieldName: string
+    value: T
+    meta: FieldMeta
+    FieldScope?: ComponentType<Provider<any>>
+    debounce: Debounce
+    runValidation: () => void
+    getInputProps: (
+      value: Partial<HTMLProps<HTMLInputElement>> &
+        Pick<HTMLProps<HTMLInputElement>, 'onSubmit'> &
+        Pick<HTMLProps<HTMLInputElement>, 'onBlur'>
+    ) => {
+      value: T
+      onChange: ChangeEventHandler<HTMLInputElement>
+      onBlur: FocusEventHandler<HTMLInputElement>
+    } & Partial<HTMLProps<HTMLInputElement>>
+    setValue: <V = any>(
```

Comment:
Can a value _really_ be `any`? Or can it only be something that is _serializable_. The problem with `any` is that is also dictates `updater`, which now becomes:

```typescript
declare AnyValueSetter = SetValue<any> = any | ...
// AnyValueSetter  = any
```

Meaning it absolutely does nothing.

Reply:
The docs say any, but serializable definitely makes sense to me. Would that type be something like the following? Or is there a built-in?

```ts
type BaseSerializable = string | number | boolean
interface Seriealizable {
  [k: string]: BaseSerializable | BaseSerializable[] | Seriealizable
}
```

Reply:
I don't know of a built-in type. I think `type-fest` also has a `JSONSerializableThingType` so that makes me think it doesn't exist yet.

Reply:
I added Serializable types to the gist I linked above, use the value type for value setters, and inferred types for the insert and push value functions if the value is an array (things like that make me looove TypeScript 😄).